rwengine/src/audio/SoundManager.cpp

```diff
@@ -59,569 +51,202 @@ bool SoundManager::initializeOpenAL() {
-        av_free(formatContext->pb->buffer);
-        avio_context_free(&formatContext->pb);
-        avcodec_close(codecContext);
-        avcodec_free_context(&codecContext);
-        avformat_close_input(&formatContext);
-        RW_ERROR("Couldn't find parametrs for context");
-        return;
-    }
-
-    // Initialize the decoder.
-    if (avcodec_open2(codecContext, codec, nullptr) != 0) {
-        av_free(formatContext->pb->buffer);
-        avio_context_free(&formatContext->pb);
-        avcodec_close(codecContext);
-        avcodec_free_context(&codecContext);
-        avformat_close_input(&formatContext);
-        RW_ERROR("Couldn't open the audio codec context");
-        return;
-    }
-#endif
-
-    // Expose audio metadata
-    channels = static_cast<size_t>(codecContext->channels);
-    sampleRate = sdt.assetInfo.sampleRate;
-
-    // OpenAL only supports mono or stereo, so error on more than 2 channels
-    if(channels > 2) {
-        RW_ERROR("Audio has more than two channels");
-        av_free(formatContext->pb->buffer);
-        avio_context_free(&formatContext->pb);
-        av_frame_free(&frame);
-        avcodec_close(codecContext);
-        avformat_close_input(&formatContext);
-        return;
-    }
+        auto emplaced =
+            sfx.emplace(std::piecewise_construct, std::forward_as_tuple(i),
+                        std::forward_as_tuple());
+        sound = &emplaced.first->second;
 
-    // Start reading audio packets
-    AVPacket readingPacket;
-    av_init_packet(&readingPacket);
-
-#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(57,37,100)
-
-    while (av_read_frame(formatContext, &readingPacket) == 0) {
-        if (readingPacket.stream_index == audioStream->index) {
-            AVPacket decodingPacket = readingPacket;
-
-            while (decodingPacket.size > 0) {
-                // Decode audio packet
-                int gotFrame = 0;
-                int len = avcodec_decode_audio4(codecContext, frame, &gotFrame, &decodingPacket);
-
-                if (len >= 0 && gotFrame) {
-                    // Write samples to audio buffer
-
-                    for(size_t i = 0; i < static_cast<size_t>(frame->nb_samples); i++) {
-                        // Interleave left/right channels
-                        for(size_t channel = 0; channel < channels; channel++) {
-                            int16_t sample = reinterpret_cast<int16_t *>(frame->data[channel])[i];
-                            data.push_back(sample);
-                        }
-                    }
-
-                    decodingPacket.size -= len;
-                    decodingPacket.data += len;
-                }
-                else {
-                    decodingPacket.size = 0;
-                    decodingPacket.data = nullptr;
-                }
-            }
-        }
-        av_free_packet(&readingPacket);
+        sound->source = std::make_shared<SoundSource>();
+        sound->source->loadSfx(path, i);
```

Comment:
``loadSfx`` has too much responsibility.

Since it's called in a loop here it would make more sense to create a ``LoaderSDT`` here and pass it in.